server/app/middlewares/websocket_backend.rb

```diff
@@ -67,29 +75,91 @@ def call(env)
     end
   end
 
-  ##
-  # On websocket connection open
-  #
-  # @param [Faye::WebSocket] ws
-  # @param [Rack::Request] req
-  def on_open(ws, req)
+  # @param node_id [String] request header Kontena-Node-Id
+  # @param grid_token [String] request header Kontena-Grid-Token
+  # @param init_attrs [Hash] initialize attributes on new node
+  # @raise [CloseError]
+  # @return [HostNode] with node_id set
+  def find_node_by_grid_token(node_id, grid_token, init_attrs)
     # check grid
-    grid = Grid.find_by(token: req.env['HTTP_KONTENA_GRID_TOKEN'].to_s)
-    unless grid
-      logger.error 'invalid grid token, closing connection'
-      ws.close(4001, "Invalid grid token")
-      return
+    grid = Grid.find_by(token: grid_token.to_s)
+
+    raise CloseError.new(4001), "Invalid grid token" unless grid
+
+    node = grid.host_nodes.find_by(node_id: node_id)
+
+    if node
+      logger.debug "node #{node} connected using grid token"
+    else
+      node = grid.host_nodes.create!(node_id: node_id, **init_attrs)
+
+      logger.info "new node #{node} connected using grid token"
+    end
+
+    return node
+  end
+
+  # @param node_id [String] request header Kontena-Node-Id
+  # @param node_token [String] request header Kontena-Node-Token
+  # @param init_attrs [Hash] initialize attributes on new node
+  # @raise [CloseError]
+  # @return [HostNode] with node_id set
+  def find_node_by_node_token(node_id, node_token, init_attrs)
+    node = HostNode.find_by(token: node_token.to_s)
```

Comment:
IMHO we should keep node_id unique for now. That said it would make sense to add that unique index there to block possible race conditions. Name should be also unique within a grid.